README.md

```diff
@@ -89,7 +89,9 @@ pip install git+https://github.com/Storia-AI/sage.git@main
     export PINECONE_INDEX_NAME=...
     ```
 
-3. For reranking, we support <a href="https://developer.nvidia.com/blog/enhancing-rag-pipelines-with-re-ranking/">NVIDIA</a>, <a href="https://docs.voyageai.com/docs/reranker">Voyage</a>, <a href="https://cohere.com/rerank">Cohere</a>, and <a href="https://jina.ai/reranker/">Jina</a>. According to [our experiments](benchmark/retrieval/README.md), NVIDIA performs best. Export the API key of the desired provider:
+3. For reranking, we support <a href="https://developer.nvidia.com/blog/enhancing-rag-pipelines-with-re-ranking/">NVIDIA</a>, <a href="https://docs.voyageai.com/docs/reranker">Voyage</a>, <a href="https://cohere.com/rerank">Cohere</a>, and <a href="https://jina.ai/reranker/">Jina</a>. According to [our experiments](benchmark/retrieval/README.md), NVIDIA performs best. Note: for NVIDIA you should use the `nvidia/nv-rerankqa-mistral-4b-v3` reranker. 
```

Comment:
It already defaults to this reranker when using NVIDIA [here](https://github.com/Storia-AI/sage/blob/40357dc0b8359063ffa46255c54d96038c0ccecf/sage/reranker.py#L38).

I think we should keep the cognitive load to a minimum in the README. Plus for the sake of consistency, if we are to specify the NVIDIA default, then we should be specifying the default for all of them.

Reply:
The confusion comes from the fact that when getting an nvidia API key to use a reranker there are multiple options to choose from in their web portal. So I want to call out which of the options to generate an API key for on the portal.

Reply:
OK then can you be more explicit about it?
Something like [here](https://docs.nvidia.com/nim/large-language-models/latest/getting-started.html#generate-an-api-key) are the instructions for how to get an API key. Note API keys are model-specific. We recommend using `nvidia/nv-rerankqa-mistral-4b-v3`.

Reply:
> The confusion comes from the fact that when getting an nvidia API key to use a reranker there are multiple options to choose from in their web portal. So I want to call out which of the options to generate an API key for on the portal.

Hello, thanks for the update. I am also trying to test Sage—on a Windows machine with Python 3.11—and wasn't sure which Nvidia API to use, but since there were only two options at [Nvidia Model rerank](https://build.nvidia.com/nvidia/nv-rerankqa-mistral-4b-v3?snippet_tab=Python), I used the latest model (`rerank-qa-mistral-4b`). Afterwards, I ran `sage-index huggingface/transformers --commit-hash=7bb1c99800d235791dace10305731f377db8077b --mode=remote
`, with these keys:
```sh
export OPENAI_API_KEY=sk-proj...
export PINECONE_API_KEY=1a...
export NVIDIA_API_KEY=nvapi-...
```
And this is the error I am getting:
```sh
WARNING:root:Unable to decode file repos\huggingface/transformers\tests\utils\test_tokenization_utils.py. Skipping.
WARNING:root:Unable to decode file repos\huggingface/transformers\utils\check_copies.py. Skipping.
WARNING:root:Unable to decode file repos\huggingface/transformers\utils\check_table.py. Skipping.
WARNING:root:Unable to decode file repos\huggingface/transformers\utils\notification_service.py. Skipping.
[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\Admin\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping tokenizers\punkt.zip.
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\Admin\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping corpora\stopwords.zip.
  0%|          | 0/2696 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "SAGE-TEST\my-venv\Scripts\sage-index.exe\__main__.py", line 7, in <module>
  File "SAGE-TEST\my-venv\Lib\site-packages\sage\index.py", line 93, in main
    repo_vector_store = build_vector_store_from_args(args, repo_manager)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "SAGE-TEST\my-venv\Lib\site-packages\sage\vector_store.py", line 182, in build_vector_store_from_args
    bm25_encoder.fit(corpus)
  File "SAGE-TEST\my-venv\Lib\site-packages\pinecone_text\sparse\bm25_encoder.py", line 87, in fit
    indices, tf = self._tf(doc)
                  ^^^^^^^^^^^^^
  File "SAGE-TEST\my-venv\Lib\site-packages\pinecone_text\sparse\bm25_encoder.py", line 281, in _tf
    counts = Counter((self._hash_text(token) for token in self._tokenizer(text)))
                                                          ^^^^^^^^^^^^^^^^^^^^^
  File "SAGE-TEST\my-venv\Lib\site-packages\pinecone_text\sparse\bm25_tokenizer.py", line 47, in __call__
    tokens = word_tokenize(text)
             ^^^^^^^^^^^^^^^^^^^
  File "SAGE-TEST\my-venv\Lib\site-packages\nltk\tokenize\__init__.py", line 142, in word_tokenize
    sentences = [text] if preserve_line else sent_tokenize(text, language)
                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "SAGE-TEST\my-venv\Lib\site-packages\nltk\tokenize\__init__.py", line 119, in sent_tokenize
    tokenizer = _get_punkt_tokenizer(language)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "SAGE-TEST\my-venv\Lib\site-packages\nltk\tokenize\__init__.py", line 105, in _get_punkt_tokenizer
    return PunktTokenizer(language)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "SAGE-TEST\my-venv\Lib\site-packages\nltk\tokenize\punkt.py", line 1744, in __init__
    self.load_lang(lang)
  File "SAGE-TEST\my-venv\Lib\site-packages\nltk\tokenize\punkt.py", line 1749, in load_lang
    lang_dir = find(f"tokenizers/punkt_tab/{lang}/")
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "SAGE-TEST\my-venv\Lib\site-packages\nltk\data.py", line 579, in find
    raise LookupError(resource_not_found)
LookupError:
**********************************************************************
  Resource punkt_tab not found.
  Please use the NLTK Downloader to obtain the resource:

  >>> import nltk
  >>> nltk.download('punkt_tab')

  For more information see: https://www.nltk.org/data.html

  Attempted to load tokenizers/punkt_tab/english/

  Searched in:
    - 'C:\\Users\\Admin/nltk_data'
    - 'D:\\SAGE-TEST\\my-venv\\nltk_data'
    - 'D:\\SAGE-TEST\\my-venv\\share\\nltk_data'
    - 'D:\\SAGE-TEST\\my-venv\\lib\\nltk_data'
    - 'C:\\Users\\Admin\\AppData\\Roaming\\nltk_data'
    - 'C:\\nltk_data'
    - 'D:\\nltk_data'
    - 'E:\\nltk_data'
```
I also want to use my own Nvidia GPU instead of NVIDIA NIM, so I am looking forward to exploring LlamaIndex since it supports `nv-rerankqa-mistral-4b-v3`

Reply:
hey @jatolentino thanks for the note. have you tried running this command with a version of sage consistent with this PR? this should fix the nltk issue

Reply:
@mihail911 I saw you merged this PR with main. So, I tried to create a new environment, and install sage again with `pip install git+https://github.com/Storia-AI/sage.git@main`, but it gave the same error. I also tried with pipx and again same error. Then, I decided to switch to a Linux OS with python 3.10.12, and after exporting the keys, I did:
```sh
python3 -m venv my-venv
source my-venv/bin/activate
pip install git+https://github.com/Storia-AI/sage.git@main
sage-index huggingface/transformers --commit-hash=7bb1c99800d235791dace10305731f377db8077b --mode=remote
```
But I continue getting the same error:

```sh
INFO:h
  File "/home/SAGE-TEST/Desktop/storia/my-venv/bin/sage-index", line 8, in <module>
    sys.exit(main())
  File "/home/SAGE-TEST/Desktop/storia/my-venv/lib/python3.10/site-packages/sage/index.py", line 97, in main
    repo_vector_store = build_vector_store_from_args(args, repo_manager)
  File "/home/SAGE-TEST/Desktop/storia/my-venv/lib/python3.10/site-packages/sage/vector_store.py", line 197, in build_vector_store_from_args
    bm25_encoder.fit(corpus)
  File "/home/SAGE-TEST/Desktop/storia/my-venv/lib/python3.10/site-packages/pinecone_text/sparse/bm25_encoder.py", line 87, in fit
    indices, tf = self._tf(doc)
  File "/home/SAGE-TEST/Desktop/storia/my-venv/lib/python3.10/site-packages/pinecone_text/sparse/bm25_encoder.py", line 281, in _tf
    counts = Counter((self._hash_text(token) for token in self._tokenizer(text)))
  File "/home/SAGE-TEST/Desktop/storia/my-venv/lib/python3.10/site-packages/pinecone_text/sparse/bm25_tokenizer.py", line 47, in __call__
    tokens = word_tokenize(text)
  File "/home/SAGE-TEST/Desktop/storia/my-venv/lib/python3.10/site-packages/nltk/tokenize/__init__.py", line 142, in word_tokenize
    sentences = [text] if preserve_line else sent_tokenize(text, language)
  File "/home/SAGE-TEST/Desktop/storia/my-venv/lib/python3.10/site-packages/nltk/tokenize/__init__.py", line 119, in sent_tokenize
    tokenizer = _get_punkt_tokenizer(language)
  File "/home/SAGE-TEST/Desktop/storia/my-venv/lib/python3.10/site-packages/nltk/tokenize/__init__.py", line 105, in _get_punkt_tokenizer
    return PunktTokenizer(language)
  File "/home/SAGE-TEST/Desktop/storia/my-venv/lib/python3.10/site-packages/nltk/tokenize/punkt.py", line 1744, in __init__
    self.load_lang(lang)
  File "/home/SAGE-TEST/Desktop/storia/my-venv/lib/python3.10/site-packages/nltk/tokenize/punkt.py", line 1749, in load_lang
    lang_dir = find(f"tokenizers/punkt_tab/{lang}/")
  File "/home/SAGE-TEST/Desktop/storia/my-venv/lib/python3.10/site-packages/nltk/data.py", line 579, in find
    raise LookupError(resource_not_found)
LookupError: 
**********************************************************************
  Resource punkt_tab not found.
  Please use the NLTK Downloader to obtain the resource:

  >>> import nltk
  >>> nltk.download('punkt_tab')
  
  For more information see: https://www.nltk.org/data.html

  Attempted to load tokenizers/punkt_tab/english/

  Searched in:
    - '/home/SAGE-TEST/nltk_data'
    - '/home/SAGE-TEST/Desktop/storia/my-venv/nltk_data'
    - '/home/SAGE-TEST/Desktop/storia/my-venv/share/nltk_data'
    - '/home/SAGE-TEST/Desktop/storia/my-venv/lib/nltk_data'
    - '/usr/share/nltk_data'
    - '/usr/local/share/nltk_data'
    - '/usr/lib/nltk_data'
    - '/usr/local/lib/nltk_data'
```

It is worth to mention that in Windows and Linux, I have the latest Ollama version installed.

Reply:
I found a workaround that fixes the error for both Windows and Linux and which I am describing at the end of this post. Seems that the `is_punkt_downloaded()` function in `sage/vector_store.py` triggers the exception: `Resource punkt_tab not found`. So I tried the following that failed:

- Modified the `is_punkt_downloaded()` like:
  ```python
  def is_punkt_downloaded():
      try:
          find('tokenizers/punkt')
          find('tokenizers/punkt_tab')
          return True
      except LookupError:
          return False
  ```
  
  ```python
  class PineconeVectorStore(VectorStore):
      def __init__(self, index_name: str, dimension: int, alpha: float, bm25_cache: Optional[str] = None):
          self.index_name = index_name
          self.dimension = dimension
          self.client = Pinecone()
          self.alpha = alpha
          if alpha < 1.0:
              if bm25_cache and os.path.exists(bm25_cache):
                  logging.info("Loading BM25 encoder from cache.")
                  # We need nltk tokenizers for bm25 tokenization
                  if is_punkt_downloaded():
                      print("punkt is already downloaded")
                  else:
                      print("punkt is not downloaded")
                      # Downloading it
                      nltk.download('punkt')
                      nltk.download('punkt_tab')
  :
  :
  ```

So, what I did was cloning the sage repo and modify the `vector_store.py` file. Within this file, I imported the `nltk` package and downloaded `punkt` and `punkt_tab` like this:
```python
import nltk
import logging
import os
:
:
def build_vector_store_from_args(args: dict, data_manager: Optional[DataManager] = None) -> VectorStore:
    """Builds a vector store from the given command-line arguments.

    When `data_manager` is specified and hybrid retrieval is requested, we'll use it to fit a BM25 encoder on the corpus
    of documents.
    """
    if args.vector_store_provider == "pinecone":
        bm25_cache = os.path.join(".bm25_cache", args.index_namespace, "bm25_encoder.json")
        nltk.download('punkt')
        nltk.download('punkt_tab')
        if not os.path.exists(bm25_cache) and data_manager:
            logging.info("Fitting BM25 encoder on the corpus...")
            corpus = [content for content, _ in data_manager.walk()]
:
:
```
Then, I installed the updated Sage repo with:
```ssh
pip install .
```
Once the package was updated, I ran:
```ssh
sage-index huggingface/transformers --commit-hash=7bb1c99800d235791dace10305731f377db8077b --mode=remote
```
And this time, the process completed successfully. If this approach or a similar solution is implemented in the `main` branch, it will be possible to do a clean installation of Sage, and run the `sage-index` command without any issues on both Windows or Linux.

Reply:
nice thanks for finding that fix @jatolentino!